#Rscript generated by Benjamin B. Currall
#Most recent update 3/29/16

################################################################
#1.1 install packages
source("http://bioconductor.org/biocLite.R")
biocLite(c("AnnotationDbi", "impute", "GO.db", "preprocessCore","org.Hs.eg.db")) 
install.packages(c("WGCNA","gplots","matrixStats","reshape","reshape2","splitstackshape","XML","plyr","rms"))
install.packages("WGCNA")
install.packages("rjson")
install.packages("ggplot2")
install.packages("plyr")

################################################################
#1.2 load libraries
library(Biobase)
library(foreign)
library(impute)
library(WGCNA)
library(org.Hs.eg.db)
library(GEOquery)
library(limma)
library(gplots)
library(preprocessCore)
library(matrixStats)
library(reshape)
library(reshape2)
library(splitstackshape)
library(XML)
library(RCurl)
library(plyr)
library(survival)
library(rms)
library(rjson)
library(ggplot2)
library(dplyr)

###1.3 set total environement drive
setwd("C:/Users/bcurrall/Desktop/DGAP/DGAP056/PCa Transcription/GEO/GSEs/")
wd <- getwd()
ddir <- file.path(wd,"data")
dir.create(ddir)
setwd(ddir)

###############################################################
###2.1 setup list of studies and gpls of interest

# ##GSE62872 - 2014 Penney
# study <-"GSE62872"
# gpl <- "GPL19370"

##GSE21032 - 2010 Taylor: large dataset so use server or local machine with lots of RAM
#study <-"GSE21032"
#gpl <- "GPL10264"

##GSE35988 - 2012 Tomlins/Grasso
# study <-"GSE35988"
# gpl <- "GPL6480"

# ##GSE3933
# #GPL2695
# study <-"GSE3933"
# gpl <- "GPL2695","GPL3044","GPL3289"


GSE62872.list<-list(GPL19370=list(GPLdata=matrix(),GPLheaders=matrix()))
GSE21032.list<-list(GPL10264=list(GPLdata=matrix(),GPLheaders=matrix()))
GSE35988.list<-list(GPL6480=list(GPLdata=matrix(),GPLheaders=matrix()))
GSE3933.list<-list(GPL2695=list(GPLdata=matrix(),GPLheaders=matrix()),GPL3044=list(GPLdata=matrix(),GPLheaders=matrix()),GPL3289=list(GPLdata=matrix(),GPLheaders=matrix()))

study.list<-list(GSE62872=GSE62872.list, GSE21032=GSE21032.list, GSE35988=GSE35988.list,GSE3933=GSE3933.list)


###############################################################
### 2.2 get and save ExpressionSet data from GEO
for(i in 1:length(study.list))
{
  for(j in 1:length(study.list[[i]]))
  {
    study<-names(study.list[i])
    gpl<-names(study.list[[i]][j])
    gset <- getGEO(study, GSEMatrix =TRUE, destdir = ddir)
    if (length(gset) > 1) idx <- grep(gpl, attr(gset, "names")) else idx <- 1
    gset <- gset[[idx]]
    prefix<-paste0(study,"_",gpl)
    gset.name <-paste0(prefix,"_gset.Rda")
    save(gset,file=gset.name)
  }
}

# save Rda if needed
study_ex_name<-paste0("study_ex_",Sys.Date(),".Rda")
save(study.ex,file=study_ex_name)

################################################################
### 2.3 get and save platform ProbeIDs from GEO
#initiate clinical data
study.ProbeIDs<-study.list
RecoveredNames<-read.table("recoverednames.txt", stringsAsFactors=FALSE)
##OR load from previous save
#load("study_ProbeIDs_2015-08-26.Rda") 


for(i in 1:length(study.list))
{
  for(j in 1:length(study.list[[i]]))
  {
    #get study info
    study<-names(study.list[i])
    gpl<-names(study.list[[i]][j])
    prefix<-paste0(study,"-",gpl)
    
    #download GEO GPL gene info
    GBurl<-paste0("http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=", gpl, "&view=data")  
    download.file(GBurl, "GBlist.txt")
    GBList_df<-data.frame(pos = 0, ProbeID = 0, RefSeq_ID=0, GB_Acc = 0, NCBI_ID = 0, Sym = 0)
    
    #setup download aluni gene info
    annotfile<-paste0(prefix,".annot.gz")
    platform_url<-paste0("ftp://ailun.stanford.edu/ailun/annotation/geo/",gpl,".annot.gz")  
    
    if(gpl == "GPL19370")
    {
      #modify GEO GPL Annotations
      GBList<-read.delim("GBlist.txt", skip = 21, header = TRUE, stringsAsFactors = FALSE)
      colnames(GBList)<-c("ProbeID","Sym")
      GBList$Sym<-gsub(".*=","",GBList$Sym)
      GBList$Sym<-gsub("\\[.*","",GBList$Sym)
      GBList$pos<-rownames(GBList)
      GBList<-GBList[1:which(grepl("<br>",GBList$ProbeID))-1,]
      
      #merge custom ID List
      ReqCol<-c("Sym", "NCBI_Sym")
      RecoveredNamesS<-RecoveredNames[,ReqCol]
      GBList <- merge(GBList, RecoveredNamesS, by="Sym", all.x = TRUE)
      
      #download and merge aluni
      #no aluni data for file
      
      #convert Aliases to Official NCBI_Symbol
      GBList$NCBI_Symbol2<-alias2SymbolTable(GBList$ProbeID)
      GBList$NCBI_Sym[which(!is.na(GBList$NCBI_Symbol2))]<-GBList$NCBI_Symbol2[which(!is.na(GBList$NCBI_Symbol2))]
      GBList$NCBI_Symbol2<-NULL
      
      #Get and merge NCBI_ID based on NCBI_Symbol
      ID.r <- unique(as.vector(as.matrix(GBList$NCBI_Sym)))
      ID.r <- ID.r[ID.r %in% ls(org.Hs.egALIAS2EG)]
      NCBI_ID <-unlist(mget(ID.r,org.Hs.egALIAS2EG))
      NCBI_Sym <- unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
      genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
      genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_Sym)),]
      GBList <- merge(GBList, genenames_NCBI, by="NCBI_Sym", all.x = TRUE)
      GBList$NCBI_ID<-as.character(GBList$NCBI_ID)
      GBList$NCBI_ID[which(is.na(GBList$NCBI_ID))]<-""
      
      #Add AnyID
      GBList$AnyID<-GBList$NCBI_ID
      GBList$AnyID[which(GBList$AnyID=="")]<-GBList$Sym[which(GBList$AnyID=="")]
      
      #add empty data columns
      emptycolumns<-c("GB_Acc","RefSeq_ID")
      GBList[,emptycolumns]<-NA
      
    }else if (gpl == "GPL10264")
    {
      #modify GEO GPL Annotations
      GBList<-read.delim("GBlist.txt", skip = 21, header = TRUE, stringsAsFactors = FALSE)
      colnames(GBList)<-c("ProbeID","RefSeq_ID")
      GBList$RefSeq_ID<-gsub(".*=","",GBList$RefSeq_ID)
      GBList$RefSeq_ID<-gsub(">.*","",GBList$RefSeq_ID)
      GBList$pos<-rownames(GBList)
      GBList<-GBList[1:which(grepl("<br>",GBList$ProbeID))-1,]
      
      #Get and merge NCBI_ID and NCBI_Sym based on RefSeq_ID
      RefSeq_ID <- unique(as.vector(as.matrix(GBList$RefSeq_ID)))
      RefSeq_ID <- RefSeq_ID[RefSeq_ID %in% ls(org.Hs.egREFSEQ2EG)]
      NCBI_ID <-unlist(mget(RefSeq_ID,org.Hs.egREFSEQ2EG))
      geneIDs_NCBI<-data.frame(RefSeq_ID,NCBI_ID)
      genenames_NCBI<-geneIDs_NCBI[which(!duplicated(geneIDs_NCBI$RefSeq_ID)),]
      GBList <- merge(GBList, geneIDs_NCBI, by="RefSeq_ID", all.x = TRUE)
      
      #retrieve gene IDS from ailun
      download.file(platform_url, annotfile)
      ncbifd <- read.csv(gzfile(annotfile),sep="\t", header=FALSE, stringsAsFactors=FALSE)
      colnames(ncbifd)<-c("ProbeID","NCBI_ID_ailun","NCBI_Symbol","Gene_Desc")
      ncbifd$NCBI_Symbol<-NULL
      ncbifd$Gene_Desc<-NULL
      ncbifd<-ncbifd[which(!duplicated(ncbifd$ProbeID)),]
      GBList <- merge(GBList, ncbifd, by="ProbeID", all.x = TRUE)
      
      #add ailun Gene IDS to RefSeq generated GeneIDs
      GBList$NCBI_ID<-as.character(GBList$NCBI_ID)
      GBList$NCBI_ID[which(is.na(GBList$NCBI_ID))]<-GBList$NCBI_ID_ailun[which(is.na(GBList$NCBI_ID))]
      GBList$NCBI_ID_ailun<-NULL
      
      #Get and merge NCBI_Symbol based on NCBI_ID
      NCBI_ID <- unique(as.vector(as.matrix(GBList$NCBI_ID)))
      NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
      NCBI_Sym <- unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
      genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
      genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_ID)),]
      GBList <- merge(GBList, genenames_NCBI, by="NCBI_ID", all.x = TRUE)
      
      #Add AnyID
      GBList$AnyID<-GBList$NCBI_ID
      GBList$AnyID[which(is.na(GBList$AnyID))]<-GBList$RefSeq_ID[which(is.na(GBList$AnyID))]
      
      #add empty data columns
      emptycolumns<-c("Sym","GB_Acc")
      GBList[,emptycolumns]<-NA
      
    }else if(gpl == "GPL6480")
    {
      GBList<-read.delim("GBlist.txt", skip = 36, header = TRUE, stringsAsFactors = FALSE)
      colnames(GBList)<-c("ProbeID","ProbePos","Cntrl","RefSeq_ID","GB_Acc","NCBI_ID","Sym","Name", "UnigeneID","EnsemblID","TIGRID","Accession","Chr","Band","Desc","GOID", "Seq")
      GBList$RefSeq_ID<-gsub(".*=","",GBList$RefSeq_ID)
      GBList$RefSeq_ID<-gsub(">.*","",GBList$RefSeq_ID)
      GBList$GB_Acc<-gsub(".*=","",GBList$GB_Acc)
      GBList$GB_Acc<-gsub(">.*","",GBList$GB_Acc)
      GBList$NCBI_ID<-gsub(".*gene/","",GBList$NCBI_ID)
      GBList$NCBI_ID<-gsub(">.*","",GBList$NCBI_ID)
      GBList$RefSeq_ID[which(GBList$RefSeq_ID=="")]<-GBList$GB_Acc[which(GBList$RefSeq_ID=="")]
      GBList$GB_Acc[grep("_",GBList$GB_Acc)]<-""
      GBList$RefSeq_ID[!grepl("_",GBList$RefSeq_ID)]<-""
      GBList$pos<-rownames(GBList)
      GBList<-GBList[1:which(grepl("<br>",GBList$ProbeID))-1,]
      
      #retrieve gene IDS from ailun
      download.file(platform_url, annotfile)
      ncbifd <- read.csv(gzfile(annotfile),sep="\t", header=FALSE, stringsAsFactors=FALSE)
      colnames(ncbifd)<-c("ProbeID","NCBI_ID_ailun","NCBI_Symbol","Gene_Desc")
      ncbifd$NCBI_Symbol<-NULL
      ncbifd$Gene_Desc<-NULL
      ncbifd<-ncbifd[which(!duplicated(ncbifd$ProbeID)),]
      GBList <- merge(GBList, ncbifd, by="ProbeID", all.x = TRUE)
      
      #add ailun Gene IDS to RefSeq generated GeneIDs
      GBList$NCBI_ID<-as.character(GBList$NCBI_ID)
      GBList$NCBI_ID[which(GBList$NCBI_ID=="")]<-GBList$NCBI_ID_ailun[which(GBList$NCBI_ID=="")]
      GBList$NCBI_ID_ailun<-NULL
      
      #get and merge files from biodb walk data
      
      #use JSON retrieval - not working
      
      ##manually retrieve
      GBfile<-paste0(prefix,"_GB_Acc.txt")
      #GB_Acc<-unique(GBList$GB_Acc)
      #write.table(GB_Acc, file = GBfile, row.names = FALSE)
      ##submit table to biodb and copy results into GBfile
      GB_AccF<-read.delim(GBfile, stringsAsFactors = FALSE)
      GB_AccF$X<-NULL
      GB_AccF$UniGene.ID<-NULL
      colnames(GB_AccF)<-c("GB_Acc","NCBI_ID_biodb")
      GB_AccF<-GB_AccF[which(!duplicated(GB_AccF$GB_Acc)),]
      GBList<-join(GBList,GB_AccF,by = "GB_Acc", type = "left", match = "all")
      GBList$NCBI_ID[which(GBList$NCBI_ID=="")]<-GBList$NCBI_ID_biodb[which(GBList$NCBI_ID=="")]
      GBList$NCBI_ID_biodb<-NULL
      
      #Get and merge NCBI_Symbol based on NCBI_ID
      NCBI_ID <- unique(as.vector(as.matrix(GBList$NCBI_ID)))
      NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
      NCBI_Sym <- unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
      genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
      genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_ID)),]
      GBList <- merge(GBList, genenames_NCBI, by="NCBI_ID", all.x = TRUE)
      
      #Add AnyID
      GBList$AnyID<-GBList$NCBI_ID
      GBList$AnyID[which(is.na(GBList$AnyID))]<-GBList$RefSeq_ID[which(is.na(GBList$AnyID))]
      GBList$AnyID[which(GBList$AnyID=="")]<-NA
      GBList$AnyID[which(is.na(GBList$AnyID))]<-GBList$GB_Acc[which(is.na(GBList$AnyID))]
      GBList$AnyID[which(GBList$AnyID=="")]<-NA
      GBList$AnyID[which(is.na(GBList$AnyID))]<-GBList$ProbeID[which(is.na(GBList$AnyID))]
      
    }else if (gpl == "GPL2695"||gpl == "GPL3044"||gpl == "GPL3289")
    {
      GBList<-read.delim("GBlist.txt", skip = 29, header = TRUE, stringsAsFactors = FALSE)
      colnames(GBList)<-c("ProbeID","MCol","MRow","Col","Row","CloneID","SpotID","Polymer", "Type","GB_Acc")
      GBList$GB_Acc<-gsub(".*=","",GBList$GB_Acc)
      GBList$GB_Acc<-gsub("></a>","",GBList$GB_Acc)
      GBList$GB_Acc<-gsub("</a>","",GBList$GB_Acc)
      GBList$GB_Acc<-gsub(">",",",GBList$GB_Acc)
      GBList$pos<-rownames(GBList)
      GBList<-GBList[1:which(grepl("<br>",GBList$ProbeID))-1,]
      
      #Get and merge NCBI_Symbol based on NCBI_ID
      GBList$GB_AccRed1<-gsub(",.*","",GBList$GB_Acc)
      GBList$GB_AccRed2<-gsub(".*,","",GBList$GB_Acc)
      
      #save(GBList, file = "GBList.Rda")
      
# ##JSON retreival - not working     
#       biodbneturl<-"http://biodbnet.abcc.ncifcrf.gov/webServices/rest.php/biodbnetRestApi.json?"
#       method<-"dbwalk"
#       path <-"GenBankNucleotideAccession->unigeneid->geneid"
#       taxonID <- "9606"
#       GBAccResults<-list()
#       for (k in 1:2)
#       {
#         GBAccF<-list
#         GBRecord<-GBList[11+k]
#         for(l in 1:ceiling(nrow(GBRecord)/900))
#         {
#           lb<-(l*900)-899
#           if(l==ceiling(nrow(GBRecord)/900))
#             {lt<-ceiling(length(GBRecord))}else
#             {lt<-l*900}
#           input <-paste(GBRecord[lb:lt,1],collapse = ",")
#           biodbnetCall<-paste0(biodbneturl,"method=",method,"&inputValues=",input,"&dbPath=",path,"&taxonID=",taxonID,"&format=row")
#           response<-getURL(biodbnetCall)
#           resp_np<-fromJSON(response)
#           GBAccF<-c(GBAccF,resp_np)
#         }
#         GBAccResults[[i]]<-GBAccF
#       }
#       
#       GBAccF<-GBAccResults[[1]] 
#       GBAccF_UL<-unlist(GBAccF)
#       GBAccF_UL<-GBAccF_UL[-1]
#       GBAcc_df<-data.frame(matrix(GBAccF_UL, ncol = 3, byrow = TRUE))
#       colnames(GBAcc_df)<-c("GB_Acc","UniGeneID","GeneID2")
#       GBAcc_df$GeneID2<-gsub("/.*","",GBAcc_df$GeneID2)
#       GBList<-cbind(GBList,GBAcc_df$GeneID2)
#       colnames(GBList[14])<-NULL
#       GBList$[which(GBList$RefSeq_ID=="")]<-GBList$GB_Acc[which(GBList$RefSeq_ID=="")]
      
      
      ##manually retrieve
#       GBfile1<-paste0(prefix,"_GB_AccRed1.txt")
#       GB_Acc<-unique(GBList$GB_AccRed1)
#       write.table(GB_Acc, file = GBfile1, row.names = FALSE)
#       GBfile2<-paste0(prefix,"_GB_AccRed2.txt")
#       GB_Acc<-unique(GBList$GB_AccRed2)
#       write.table(GB_Acc, file = GBfile2, row.names = FALSE)
      
      ##submit table to biodb and copy results into GBfile
      
      #First GB_Acc Number
      GBfile1<-paste0(prefix,"_GB_AccRed1.txt")
      GBAccF1<-read.delim(GBfile1, stringsAsFactors = FALSE)
      GBAccF1$X<-NULL
      GBAccF1$UniGene.ID<-NULL
      colnames(GBAccF1)<-c("GB_AccRed1","NCBI_ID1")
      GBAccF1<-GBAccF1[which(!duplicated(GBAccF1$GB_AccRed1)),]
      GBList<-join(GBList,GBAccF1,by = "GB_AccRed1", type = "left", match = "all")
      #Last GB_Acc Number
      GBfile2<-paste0(prefix,"_GB_AccRed2.txt")
      GBAccF2<-read.delim(GBfile2, stringsAsFactors = FALSE)
      GBAccF2$X<-NULL
      GBAccF2$UniGene.ID<-NULL
      colnames(GBAccF2)<-c("GB_AccRed2","NCBI_ID2")
      GBAccF2<-GBAccF2[which(!duplicated(GBAccF2$GB_AccRed2)),]
      GBList<-join(GBList,GBAccF2,by = "GB_AccRed2", type = "left", match = "all")
      
      #Combine NCBI Gene IDs
      GBList$NCBI_ID1<-gsub("-","",GBList$NCBI_ID1)
      GBList$NCBI_ID1[is.na(GBList$NCBI_ID1)]<-""
      GBList$NCBI_ID2<-gsub("-","",GBList$NCBI_ID2)
      GBList$NCBI_ID2[is.na(GBList$NCBI_ID2)]<-""
      GBList$NCBI_ID1[which(GBList$NCBI_ID1=="")]<-GBList$NCBI_ID2[which(GBList$NCBI_ID1=="")]
      GBList$NCBI_ID1<-gsub(";.*","",GBList$NCBI_ID1)
      GBList$NCBI_ID2<-NULL
      colnames(GBList)[14]<-"NCBI_ID_biodb"
      
      #retrieve gene IDS from ailun
      download.file(platform_url, annotfile)
      ncbifd <- read.csv(gzfile(annotfile),sep="\t", header=FALSE, stringsAsFactors=FALSE)
      colnames(ncbifd)<-c("ProbeID","NCBI_ID_ailun","NCBI_Symbol","Gene_Desc")
      ncbifd$NCBI_Symbol<-NULL
      ncbifd$Gene_Desc<-NULL
      ncbifd<-ncbifd[which(!duplicated(ncbifd$ProbeID)),]
      
      GBList<-merge(GBList, ncbifd, by="ProbeID", all.x=TRUE)
      GBList$NCBI_ID_ailun[which(is.na(GBList$NCBI_ID_ailun))]<-GBList$NCBI_ID_biodb[which(is.na(GBList$NCBI_ID_ailun))]
      GBList$NCBI_ID_biodb<-NULL
      colnames(GBList)[14]<-"NCBI_ID"
      
      NCBI_ID <- unique(as.vector(as.matrix(GBList$NCBI_ID)))
      NCBI_ID <- NCBI_ID[NCBI_ID %in% ls(org.Hs.egSYMBOL)]
      NCBI_Sym <- unlist(mget(NCBI_ID,org.Hs.egSYMBOL))
      genenames_NCBI<-data.frame(NCBI_ID,NCBI_Sym)
      genenames_NCBI<-genenames_NCBI[which(!duplicated(genenames_NCBI$NCBI_ID)),]
      GBList <- merge(GBList, genenames_NCBI, by="NCBI_ID", all.x = TRUE)
      
      #Add AnyID
      GBList$AnyID<-GBList$NCBI_ID
      GBList$AnyID[which(GBList$AnyID=="")]<-GBList$GB_AccRed1[which(GBList$AnyID=="")]
      GBList$AnyID[which(GBList$AnyID=="")]<-GBList$CloneID[which(GBList$AnyID=="")]
      GBList$AnyID[which(GBList$AnyID=="")]<-GBList$SpotID[which(GBList$AnyID=="")]
      
      emptycolumns<-c("Sym","RefSeq_ID")
      GBList[,emptycolumns]<-NA
    }
    GBList_df<-GBList[,c("pos","ProbeID", "Sym", "NCBI_Sym", "NCBI_ID","RefSeq_ID","GB_Acc", "AnyID")]
    GBList_df$pos<-as.numeric(as.character(GBList_df$pos))
    GBList_df <- GBList_df[order(GBList_df$pos),]
    rownames(GBList_df)<-GBList_df$pos
    study.ProbeIDs[[i]][[j]]$ProbeIds<-GBList_df
  }
}
 

# save Rda if needed
study.ProbeIDs_name<-paste0("study_ProbeIDs_",Sys.Date(),".Rda")
save(study.ProbeIDs,file=study.ProbeIDs_name)

################################################################
### 3.1 load expression data and make proper column names and setup expression table
###??? I am using the Expression set reported results (i.e. exprs() function from Biocore), 
###??? which will vary from author to author.
###??? Do we have to process from raw microchip data or using the reported results fine?


study.ex<-study.ProbeIDs
for(i in 1:length(study.list))
{
  for(j in 1:length(study.list[[i]]))
  {
    study<-names(study.list[i])
    gpl<-names(study.list[[i]][j])
    prefix<-paste0(study,"_",gpl)
    gset.name <-paste0(prefix,"_gset.Rda")
    load(gset.name)
    fvarLabels(gset)<-make.names(fvarLabels(gset))
    #log2 expression from GEO2R
    ex <- exprs(gset)
    qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
    LogC <- (qx[5] > 100) ||
      (qx[6]-qx[1] > 50 && qx[2] > 0) ||
      (qx[2] > 0 && qx[2] < 1 && qx[4] > 1 && qx[4] < 2)
    if (LogC) { ex[which(ex <= 0)] <- NaN
    exprs(gset) <- log2(ex) }
    study.ex[[i]][[j]][[1]]<-ex
    study.ex[[i]][[j]][[2]]<-matrix(colnames(study.ex[[i]][[j]][[1]]))
    colnames(study.ex[[i]][[j]][[2]])[1]<-"GM"
    
    #add subject IDs
    if(study=="GSE62872")
    {
      #add subject IDs
      sample.info<-data.frame(as.character(gset@phenoData@data$title))
      colnames(sample.info)[1]<-"ID_Raw"
      sample.info$ID_Simple<-sample.info$ID_Raw
      sample.info$ID_Simple<-gsub("Prostate_Tumor_","",sample.info$ID_Simple)
      sample.info$ID_Simple<-gsub("Normal_Prostate_","",sample.info$ID_Simple)
      
      #add tissue status and simplify to "primary", "normal", "met" or "cell
      sample.info$Status_Raw<-as.character(gset@phenoData@data$source_name_ch1)
      sample.info$Status_Simple<-sample.info$Status_Raw
      sample.info$Status_Simple[grepl("normal",sample.info$Status_Simple)]<-"normal"
      sample.info$Status_Simple[grepl("tumor",sample.info$Status_Simple)]<-"primary"
      study.ex[[i]][[j]][[2]]<-cbind(study.ex[[i]][[j]][[2]],sample.info)
    }else if (study=="GSE21032")
    {
      #add subject IDs
      sample.info<-data.frame(as.character(gset@phenoData@data$characteristics_ch1))
      colnames(sample.info)[1]<-"ID_Raw"
      sample.info$ID_Simple<-sample.info$ID_Raw
      sample.info$ID_Simple<-gsub("sample id:","",sample.info$ID_Simple)
      
      #add tissue status and simplify to "normal", "primary", "met" or "cell
      sample.info$Status_Raw<-gset@phenoData@data$characteristics_ch1.2
      sample.info$Status_Type<-gset@phenoData@data$characteristics_ch1.3
      sample.info$Status_Simple<-paste(sample.info$Status_Raw,sample.info$Status_Type)
      sample.info$Status_Simple[grepl("normal",sample.info$Status_Simple)]<-"normal"
      sample.info$Status_Simple[grepl("Primary",sample.info$Status_Simple)]<-"primary"
      sample.info$Status_Simple[grepl("Metastasis",sample.info$Status_Simple)]<-"met"
      sample.info$Status_Simple[grepl("cell",sample.info$Status_Simple)]<-"cell"
      sample.info$Status_Type<-NULL
      study.ex[[i]][[j]][[2]]<-cbind(study.ex[[i]][[j]][[2]],sample.info)
      
    }else if (study=="GSE35988")
    {
      #add subject IDs
      sample.info<-data.frame(as.character(gset@phenoData@data$title))
      colnames(sample.info)[1]<-"ID_Raw"
      sample.info$ID_Simple<-sample.info$ID_Raw
      
      #add tissue status and simplify to "normal", "primary", "met" or "cell
      sample.info$Status_Raw<-as.character(gset@phenoData@data$characteristics_ch2)
      sample.info$Status_Simple<-sample.info$Status_Raw
      sample.info$Status_Simple[grepl("benign",sample.info$Status_Simple)]<-"normal"
      sample.info$Status_Simple[grepl("localized",sample.info$Status_Simple)]<-"primary"
      sample.info$Status_Simple[grepl("metastatic",sample.info$Status_Simple)]<-"met"
      study.ex[[i]][[j]][[2]]<-cbind(study.ex[[i]][[j]][[2]],sample.info)
      
    }else if (study=="GSE3933")
    {
      #add subject IDs
      sample.info<-data.frame(as.character(gset@phenoData@data$title))
      colnames(sample.info)[1]<-"ID_Raw"
      sample.info$ID_Simple<-sample.info$ID_Raw
      sample.info$ID_Simple<-gsub("_.","",sample.info$ID_Simple)
      
      #add tissue status and simplify to "normal", "primary", "met" or "cell
      sample.info$Status_Raw<-as.character(gset@phenoData@data$title)
      sample.info$Status_Simple<-sample.info$Status_Raw
      sample.info$Status_Simple[grepl("N",sample.info$Status_Simple)]<-"normal"
      sample.info$Status_Simple[grepl("T",sample.info$Status_Simple)]<-"primary"
      sample.info$Status_Simple[grepl("L",sample.info$Status_Simple)]<-"met"
      study.ex[[i]][[j]][[2]]<-cbind(study.ex[[i]][[j]][[2]],sample.info) 
    }
  }
}
# save Rda if needed
study_ex_name<-paste0("study_ex_",Sys.Date(),".Rda")
save(study.ex,file=study_ex_name)


##########################################################################
###3.2 get uniform probe IDs and select best probe per gene - takes a long time

#prepare file for using gene ID number - uses NCBI ID
study.ex.genes<-study.ex 
#OR pull previous version
#load("study_ex_2015-08-17.Rda") 

#quantile normalization here

for(i in 1:length(study.ex))
{
  for(j in 1:length(study.ex[[i]]))
  {
    study<-names(study.ex[i])
    gpl<-names(study.ex[[i]][j])
    
    #reduce gene list to reported ex
    genenames<-study.ex.genes[[i]][[j]]$ProbeIds
    cstudy<-data.frame(study.ex.genes[[i]][[j]]$GPLdata)
    cstudy$ProbeID<-rownames(cstudy)
    cstudy<-cstudy[order(cstudy$ProbeID),]
    cstudy$ProbeID<-NULL
    ProbeToGene<-data.frame(rownames(cstudy))
    colnames(ProbeToGene)<-"ProbeID"
    ProbeToGene<-merge(ProbeToGene,genenames, by="ProbeID", all.x=TRUE)
    ProbeToGene<-ProbeToGene[order(ProbeToGene$ProbeID),]
    
    #combine probes with same AnyID - based on PMID:21816037 (Note: takes a few minutes)
    collapse.object<-collapseRows(datET=cstudy, rowGroup=ProbeToGene$AnyID, rowID=rownames(cstudy))
    redDatET<-data.frame(collapse.object$datETcollapsed)
    study.ex.genes[[i]][[j]]$GPLdata<-redDatET
  }
}

study_ex_genes_name<-paste0("study_ex_genes_",Sys.Date(),".Rda")
save(study.ex.genes,file=study_ex_genes_name)


###############################################################
### 3.3 Load and simplify clinical data - must be downloaded from individual studies

#initiate clinical data
study.cdata<-study.ex.genes

##OR load from previous save
#load("study_ex_genes_2015-09-08.Rda") 

for(i in 1:length(study.cdata))
{
  study<-names(study.cdata[i])
  cstudy.name<-paste0(study,"_ClinicalData.txt")
  if(study=="GSE62872")
  {
    survtable<-data.frame(study.ex.genes$GSE62872$GPL19370$GPLheaders$ID_Simple)
    colnames(survtable)[1]<-"ID_Simple"
    survtable$Surv_Time<-"NA"
    survtable$Surv_Event<-"NA"
    study.cdata$GSE62872$GPL19370$Cdata<-survtable
  }else if (study=="GSE21032")
  {
    cdata<-data.frame(read.delim(cstudy.name))
    survtable<-data.frame(cdata$SampleID)
    colnames(survtable)[1]<-"ID_Simple"
    survtable$ID_Simple<-as.character(survtable$ID_Simple)
    survtable$Surv_Time<-cdata$BCR_FreeTime
    survtable$Surv_Event<-cdata$BCR_Event
    survtable$Surv_Event<-gsub("NO",0,survtable$Surv_Event)
    survtable$Surv_Event<-gsub("BCR_Algorithm",1,survtable$Surv_Event)
    study.cdata$GSE21032$GPL10264$Cdata<-survtable
  }else if (study=="GSE35988")
  {
    cdata<-data.frame(read.delim(cstudy.name))
    survtable<-data.frame(cdata$Sample.Name)
    colnames(survtable)[1]<-"ID_Simple"  
    survtable$Surv_Time<-cdata$Survival.from.diagnosis..mo.8
    survtable$Surv_Event<-1
    study.cdata$GSE35988$GPL6480$Cdata<-survtable
  }else if (study=="GSE3933")
  {
    cdata<-data.frame(read.delim(cstudy.name))
    survtable<-data.frame(cdata$Sample)
    colnames(survtable)[1]<-"ID_Simple"  
    survtable$Surv_Time<-cdata$Recurrence.free.survival..months.
    survtable$Surv_Event<-cdata$Recurrence.
    study.cdata$GSE3933$GPL2695$Cdata<-survtable
    study.cdata$GSE3933$GPL3044$Cdata<-survtable
    study.cdata$GSE3933$GPL3289$Cdata<-survtable
  } 
}

study_cdata_name<-paste0("study_cdata_",Sys.Date(),".Rda")
save(study.cdata,file=study_cdata_name)

#########################################################################
###4.1 normaliziation
###??? Is there any perferred methods for normalizing accross Expression sets?

##load and initiate clinical data
#load("study_cdata_2015-08-17.Rda") 
study.z<-study.cdata

# ## log2 transform
# qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
# LogC <- (qx[5] > 100) ||
#   (qx[6]-qx[1] > 50 && qx[2] > 0) ||
#   (qx[2] > 0 && qx[2] < 1 && qx[4] > 1 && qx[4] < 2)
# if (LogC) { ex[which(ex <= 0)] <- NaN
#   ex <- log2(ex) } #not sure if this lines works
# ex.log2.name <- paste0(sprefix,"_log2.txt")
# write.table(ex, file = ex.raw.name, quote = FALSE, sep = "\t", row.names = TRUE, col.names = TRUE)

## quantile normalization
for(i in 1:length(study.z))
{
  for(j in 1:length(study.z[[i]]))
  {
    study<-names(study.z)[i]
    gpl<-names(study.z[[i]][j])
    ex<-as.matrix(study.z[[i]][[j]]$GPLdata)
    ex.qnorm<-normalize.quantiles(ex ,copy=TRUE)
    rownames(ex.qnorm)<-rownames(ex)
    colnames(ex.qnorm)<-colnames(ex)
    study.z[[i]][[j]]$GPLqnorm<-ex.qnorm
  }
}



##load and initiate clinical data
#load("study_cdata_2015-08-17.Rda") 

##z-values for quantiles
for(i in 1:length(study.z))
{
  for(j in 1:length(study.z[[i]]))
  {
    study<-names(study.z)[i]
    gpl<-names(study.z[[i]][j])
    ex<-study.z[[i]][[j]]$GPLqnorm
    colnames(ex)<-study.z[[i]][[j]]$GPLheaders$Status_Simple
    ex.normal <- as.matrix(ex[,c(grepl("normal",colnames(ex)))])
    ex.z <-(ex-rowMeans(ex.normal, na.rm = TRUE))/rowSds(ex.normal, na.rm = TRUE)
    colnames(ex.z)<-study.z[[i]][[j]]$GPLheaders$GM
    study.z[[i]][[j]]$GPLQNormZ<-ex.z
  }
}

## Z-values only
for(i in 1:length(study.z))
{
  for(j in 1:length(study.z[[i]]))
  {
    study<-names(study.z)[i]
    gpl<-names(study.z[[i]][j])
    ex<-study.z[[i]][[j]]$GPLdata
    colnames(ex)<-study.z[[i]][[j]]$GPLheaders$Status_Simple
    ex.normal <- as.matrix(ex[,c(grepl("normal",colnames(ex)))])
    ex.z <-(ex-rowMeans(ex.normal, na.rm = TRUE))/rowSds(ex.normal, na.rm = TRUE)
    colnames(ex.z)<-study.z[[i]][[j]]$GPLheaders$GM
    study.z[[i]][[j]]$ZValues<-ex.z
  }
}
study_z_name<-paste0("study_z_",Sys.Date(),".Rda")
save(study.z,file=study_z_name)

##########################################################
###4.2 combine all studies
##load and initiate clinical data
#load("study_z_2015-08-17.Rda") 

study.full<-study.z

alldataQ<-data.frame()
alldataZ<-data.frame()
alldataQZ<-data.frame()
alldataPI<-data.frame()

alldata.headers<-data.frame()
alldata.cdata<-data.frame()

for(i in 1:length(study.full))
{
  fullstudyQ<-data.frame()
  fullstudyZ<-data.frame()
  fullstudyQZ<-data.frame()
  fullstudyPI<-data.frame()
  
  fullstudy.headers<-data.frame()
  fullstudy.cdata<-data.frame()
  for(j in 1:length(study.full[[i]]))
  {
    Qdata<-data.frame(study.full[[i]][[j]]$GPLqnorm)
    Qdata$name<-rownames(Qdata)
    fullstudyQ$name<-rownames(fullstudyQ)
    fullstudyQ<-join(fullstudyQ,Qdata, by="name", type="full")
    rownames(fullstudyQ)<-fullstudyQ$name
    fullstudyQ$name<-NULL
    
    Zdata<-data.frame(study.full[[i]][[j]]$ZValues)
    Zdata$name<-rownames(Zdata)
    fullstudyZ$name<-rownames(fullstudyZ)
    fullstudyZ<-join(fullstudyZ,Zdata, by="name", type="full")
    rownames(fullstudyZ)<-fullstudyZ$name
    fullstudyZ$name<-NULL
    
    QZdata<-data.frame(study.full[[i]][[j]]$GPLQNormZ)
    QZdata$name<-rownames(QZdata)
    fullstudyQZ$name<-rownames(fullstudyQZ)
    fullstudyQZ<-join(fullstudyQZ,QZdata, by="name", type="full")
    rownames(fullstudyQZ)<-fullstudyQZ$name
    fullstudyQZ$name<-NULL
    
    ProbeIDs<-data.frame(study.full[[i]][[j]]$ProbeIds)
    if(length(fullstudyPI)<1)
    {
      fullstudyPI<-ProbeIDs
    }else
    {
      fullstudyPI<-join(fullstudyPI,ProbeIDs, by="AnyID", type="left")
    }
    
    fullstudy.headers<-rbind(fullstudy.headers,study.full[[i]][[j]]$GPLheaders)
    fullstudy.cdata<-unique(rbind(fullstudy.cdata,study.full[[i]][[j]]$Cdata))
  }
  
  study.full[[i]]$fullstudy$GPLqnorm<-fullstudyQ
  study.full[[i]]$fullstudy$ZValues<-fullstudyZ
  study.full[[i]]$fullstudy$GPLQNormZ<-fullstudyQZ
  study.full[[i]]$fullstudy$ProbeIds<-fullstudyPI
  
  study.full[[i]]$fullstudy$GPLheaders<-fullstudy.headers
  study.full[[i]]$fullstudy$Cdata<-fullstudy.cdata
  
  Qdata<-data.frame(study.full[[i]]$fullstudy$GPLqnorm)
  Qdata$name<-rownames(Qdata)
  alldataQ$name<-rownames(alldataQ)
  alldataQ<-join(alldataQ,Qdata, by="name", type="full")
  rownames(alldataQ)<-alldataQ$name
  alldataQ$name<-NULL
  
  Zdata<-data.frame(study.full[[i]]$fullstudy$ZValues)
  Zdata$name<-rownames(Zdata)
  alldataZ$name<-rownames(alldataZ)
  alldataZ<-join(alldataZ,Zdata, by="name", type="full")
  rownames(alldataZ)<-alldataZ$name
  alldataZ$name<-NULL
  
  QZdata<-data.frame(study.full[[i]]$fullstudy$GPLQNormZ)
  QZdata$name<-rownames(QZdata)
  alldataQZ$name<-rownames(alldataQZ)
  alldataQZ<-join(alldataQZ,QZdata, by="name", type="full")
  rownames(alldataQZ)<-alldataQZ$name
  alldataQZ$name<-NULL
  
  ProbeIDs<-data.frame(study.full[[i]]$fullstudy$ProbeIds)
  if(length(alldataPI)<1)
  {
    alldataPI<-ProbeIDs
  }else
  {
    alldataPI<-join(alldataPI,ProbeIDs, by="AnyID", type="left")
  }
  
  alldata.headers<-rbind(alldata.headers,study.full[[i]]$fullstudy$GPLheaders)
  alldata.cdata<-rbind(alldata.cdata,study.full[[i]]$fullstudy$Cdata)
}

study.full$alldata$fullstudy$GPLqnorm<-alldataQ
study.full$alldata$fullstudy$ZValues<-alldataZ
study.full$alldata$fullstudy$GPLQNormZ<-alldataQZ
study.full$alldata$fullstudy$ProbeIds<-alldataPI

study.full$alldata$fullstudy$GPLheaders<-alldata.headers
study.full$alldata$fullstudy$Cdata<-alldata.cdata

study_full<-paste0("study_full",Sys.Date(),".Rda")
save(study.full,file=study_full)


# ##########################################################
# ###4.3 Define Low, Normal and High Z-values for GOI
# ##load and initiate full study
# #load("study_z_2015-10-02.Rda") 
# 
# study.status<-study.full
# 
# goi<-"C2orf43"
# 
# for(i in 1:length(study.status))
# {
#   for(j in 1:length(study.status[[i]]))
#   {
#     status<-study.status[[i]][[j]]$GPLqnorm
#     colnames(status)<-study.status[[i]][[j]]$GPLheaders$Status_Simple
#     rownames(status)<-study.status[[i]][[j]]$ProbeIds$AnyID
#     z.normal <- status[,c(grepl("normal",colnames(status)))]
#     z.normal.number<-data.frame(rowSums(!is.na(z.normal)))
#     z.normal.summary <- data.frame(rowSums(z.normal>=2,na.rm = TRUE),rowSums(z.normal<=-2,na.rm = TRUE),row.names=rownames(z.normal))
#     names(z.normal.summary)<-c("z>=2","z<=-2")
#     z.normal.summary$total<-z.normal.summary$`z>=2`-z.normal.summary$`z<=-2`
#     z.normal.summary$number<-rowSums(!is.na(z.normal))
#   }
# }  
# 
# ep.qz<-study.full[[GSE]][[GPL]]$GPLQNormZ
# colnames(ep.qz)<-cstudy.headers$Status_Simple
# ep.status<-data.frame(tissue=colnames(ep.qz))
# ep.status$sig<-""
# ep.status$sig[which(ep.edges[goi,]<=zlow)]<-"low"
# ep.status$sig[which(ep.edges[goi,]>=zhigh)]<-"high"
# ep.status$sig[which(ep.edges[goi,]>zlow & ep.edges[goi,]<zhigh)]<-"normal"
# ep.status$pos<-as.numeric(as.character(rownames(ep.status)))
# 
# 
# 



###########################################################
########Section 5.0 - Data Analysis########################
###########################################################
###
load("study_full2015-10-02.Rda") 



####################################################################
##specify study
#All: GSE = "alldata", GPL = "fullstudy"
#2014 Penney:         GSE = "GSE62872", GPL = "fullstudy" or "GPL19370"
#2010 Taylor:         GSE = "GSE21032", GPL = "fullstudy" or "GPL10264"
#2012 Tomlins/Grasso: GSE = "GSE35988", GPL = "fullstudy" or "GPL6480"
#2005 LaPointe:       GSE = "GSE3933",  GPL = "fullstudy" or "GPL2695" or "GPL3044" or "GPL3289"

#load study.full (optional - required if starting analysis with new session)
#load("study_full_YYYY-MM-DD.Rda")

GSE <- "GSE3933" 
GPL <- "fullstudy"
data<-"GPLQNormZ" #either "GPLqnorm", "ZValues" or "GPLQNormZ"
cstudy.ex<-study.full[[GSE]][[GPL]][[data]]
cstudy.headers<-study.full[[GSE]][[GPL]]$GPLheaders
cstudy.cdata<-study.full[[GSE]][[GPL]]$Cdata
cstudy.ProbeID<-study.full[[GSE]][[GPL]]$ProbeIds

##specify goi
goi<-"C2orf43"

##convert NCBI_IDs to gene symbols (Optional)
ProbeIDs<-data.frame(rownames(cstudy.ex))
colnames(ProbeIDs)<-"AnyID"
ProbeIDs$cpos<-rownames(ProbeIDs)
ProbeIDs<-join(ProbeIDs,cstudy.ProbeID, by="AnyID", type="full")
ProbeIDs<-ProbeIDs[which(!duplicated(ProbeIDs$cpos)),]
ProbeIDs<-ProbeIDs[which(!is.na(ProbeIDs$cpos)),]
cstudy.ex$name<-ProbeIDs$NCBI_Sym #use any desired ID from ProbeIDs
cstudy.ex<-cstudy.ex[which(!is.na(cstudy.ex$name)),]
rownames(cstudy.ex)<-cstudy.ex$name
cstudy.ex$name<-NULL

# #use all transcripts NCBI known gene symbols
# rownames(cstudy.ex)<-make.names(SymbolsFrame$AnyName, unique = TRUE)


###########################################################
###5.1A Group by status - needed for boxplot (5.1B) and rank order (5.1C)


for(i in 1:length(study.full))
  
##Change headers to status
colnames(cstudy.ex)<-cstudy.headers$Status_Simple

##Summary of status (normal, primary and met)
z.normal <- cstudy.ex[,c(grepl("normal",colnames(cstudy.ex)))]
z.normal.number<-data.frame(rowSums(!is.na(z.normal)))
z.normal.summary <- data.frame(rowSums(z.normal>=2,na.rm = TRUE),rowSums(z.normal<=-2,na.rm = TRUE),row.names=rownames(z.normal))
names(z.normal.summary)<-c("z>=2","z<=-2")
z.normal.summary$total<-z.normal.summary$`z>=2`-z.normal.summary$`z<=-2`
z.normal.summary$number<-rowSums(!is.na(z.normal))

z.primary <- cstudy.ex[,c(grepl("primary",colnames(cstudy.ex)))]
z.primary.summary <- data.frame(rowSums(z.primary>=2,na.rm = TRUE),rowSums(z.primary<=-2,na.rm = TRUE),row.names=rownames(z.primary))
names(z.primary.summary)<-c("z>=2","z<=-2")
z.primary.summary$total<-z.primary.summary$`z>=2`-z.primary.summary$`z<=-2`
z.primary.summary$number<-rowSums(!is.na(z.primary))

if (sum(grepl("met",colnames(cstudy.ex)))>0)
{
  z.met <- cstudy.ex[,c(grepl("met",colnames(cstudy.ex)))]
  z.met.summary <- data.frame(rowSums(z.met>=2,na.rm = TRUE),rowSums(z.met<=-2,na.rm = TRUE),row.names=rownames(z.met))
  names(z.met.summary)<-c("z>=2","z<=-2")
  z.met.summary$total<-z.met.summary$`z>=2`-z.met.summary$`z<=-2` 
  z.met.summary$number<-rowSums(!is.na(z.met))
  
  z.full.summary<-data.frame(z.normal.summary$total,z.normal.summary$number, z.primary.summary$total,z.primary.summary$number,z.met.summary$total,z.met.summary$number,row.names = rownames(z.normal.summary))
  names(z.full.summary)<-c("n_diff","n_tot","p_diff","p_tot","m_diff","m_tot")
  
}else
{
  z.full.summary<-data.frame(z.normal.summary$total,z.normal.summary$number,z.primary.summary$total,z.primary.summary$number,row.names = rownames(z.normal.summary))
  names(z.full.summary)<-c("n_diff","n_tot","p_diff","p_tot")
}

###########################################################
###5.1B boxplot by goi status - Requires 5.1A

##prepare data

goi.normal<-data.frame(z.normal[goi,])
if(nrow(goi.normal)<2){goi.normal<-data.frame(t(goi.normal))}
goi.primary<-data.frame(z.primary[goi,])
if(nrow(goi.primary)<2){goi.primary<-data.frame(t(goi.primary))}
if (sum(grepl("met",colnames(cstudy.ex)))>0)
{
  goi.met<-data.frame(z.met[goi,])
  if(nrow(goi.met)<2){goi.met<-data.frame(t(goi.met))}
  goi.list<-list(goi.normal,goi.primary,goi.met)
  goi.maxrows<- max(nrow(goi.normal),nrow(goi.primary),nrow(goi.met))
  
  goi.df <- lapply(goi.list, function(x) { x[1:goi.maxrows,] })
  goi.df<-do.call(cbind, lapply(goi.df, `[`))
  colnames(goi.df)<-c("benign", "primary", "met")
  
  }else
{
  goi.list<-list(goi.normal,goi.primary)
  goi.maxrows<- max(nrow(goi.normal),nrow(goi.primary))
  
  goi.df <- lapply(goi.list, function(x) { x[1:goi.maxrows,] })
  goi.df<-do.call(cbind, lapply(goi.df, `[`))
  colnames(goi.df)<-c("benign", "primary")
}

##boxplot
par(mar=c(2,2,1,1))
boxplot(goi.df, ylim=c(-11,9))

write.table(goi.df, "cstudy_goi_df.txt")
##statistics

#t-test if only benign vs. primary
t.test(goi.df[,"benign"],goi.df[,"primary"])

#ANOVA/Summary and t-test if normal, primary and mets
goi.df.melt<-melt(goi.df)
colnames(goi.df.melt)<-c("#","status","z_values")
goi.df.av<-aov(z_values~status, data=goi.df.melt)
summary(goi.df.av)

t.test(goi.df[,"benign"],goi.df[,"primary"])
t.test(goi.df[,"benign"],goi.df[,"met"])
t.test(goi.df[,"primary"],goi.df[,"met"])

goi.bplot.name <-paste0("full","_",goi,"_bplot.txt")
write.table(goi.df, file = goi.bplot.name, quote = FALSE, sep = "\t", row.names = TRUE, col.names = TRUE)

###########################################################
###5.1B-2 boxplot of markers vs. goi

##Set Markers
#Defined in PMID:20426842
#Manually save Additional File 1 and 2 into WD as CSVs
MarkEpi<-read.csv("20426842_AF1.csv", skip = 2, header=TRUE)
MarkEpi<-MarkEpi[which(!duplicated(MarkEpi$Gene.symbol)),]
GeneConv<-AnnotationDbi::select(org.Hs.eg.db, keys=c(as.character(MarkEpi$Gene.symbol)), columns=c("ALIAS","SYMBOL", "ENTREZID"), keytype="ALIAS")
GeneConv<-GeneConv[which(!duplicated(GeneConv$ALIAS)),]
MarkEpi$Sym<-GeneConv$SYMBOL
MarkEpi<-MarkEpi[which(!is.na(MarkEpi$Sym)),]

MarkStroma<-read.csv("20426842_AF2.csv", skip = 2, header=TRUE)
MarkStroma<-MarkStroma[which(!duplicated(MarkStroma$Gene.symbol)),]
GeneConv<-AnnotationDbi::select(org.Hs.eg.db, keys=c(as.character(MarkStroma$Gene.symbol)), columns=c("ALIAS","SYMBOL", "ENTREZID"), keytype="ALIAS")
GeneConv<-GeneConv[which(!duplicated(GeneConv$ALIAS)),]
MarkStroma$Sym<-GeneConv$SYMBOL
MarkStroma<-MarkStroma[which(!is.na(MarkStroma$Sym)),]

##Prepare Data for GOI
Epi<-t(cstudy.ex[MarkEpi$Sym,])
Epi<-melt(Epi)
Epi$tissue<-"epithelia"
Stroma<-t(cstudy.ex[MarkStroma$Sym,])
Stroma<-melt(Stroma)
Stroma$tissue<-"stroma"
Goi<-t(cstudy.ex[goi,])
Goi<-melt(Goi)
Goi$tissue<-goi
Markers<-rbind(Epi,Stroma,Goi)

colnames(Markers)<-c("status","gene","z","tissue")
Markers$status<-as.character(Markers$status)
Markers<-Markers[Markers$status!="cell",]
Markers$status<-gsub("normal","benign",Markers$status)
Markers$status<-gsub("primary","tumor",Markers$status)
Markers$status<-gsub("met","tumor",Markers$status)
Markers$status<-as.factor(Markers$status)
Markers$tissue<-as.factor(Markers$tissue)
Markers<-as.data.frame(Markers)

Markers<-tbl_df(Markers)

Markers.sort<-Markers %>%
  filter(!is.na(z))%>%
  filter(status=="tumor")%>%
  group_by(gene)%>%
  summarise(zmean=mean(z))%>%
  arrange(zmean)


Markers$gene <- factor(Markers$gene, levels = Markers.sort$gene)
Markers$tissue <- factor(Markers$tissue, levels = c(goi,"stroma","epithelia"))

Markers.f.epi<-Markers %>%
  filter(!is.na(z))%>%
  filter(status=="tumor")%>%
  filter(tissue=="epithelia")%>%
  group_by(gene)
  
Markers.f.stroma<-Markers %>%
  filter(!is.na(z))%>%
  filter(status=="tumor")%>%
  filter(tissue=="stroma")%>%
  group_by(gene)

Markers.f.tissue<-Markers %>%
  filter(!is.na(z))%>%
  filter(status=="tumor")%>%
  group_by(tissue)

ggplot(Markers.f.epi, aes(status, z))+geom_boxplot()+facet_grid(tissue~gene)+ylim(-10,10)
ggplot(Markers.f.stroma, aes(status, z))+geom_boxplot()+facet_grid(tissue~gene)+ylim(-10,10)
ggplot(Markers.f.tissue, aes(status, z))+geom_boxplot()+facet_grid(.~tissue)+ylim(-5,5)

t.test(Markers$z[Markers$tissue==goi],Markers$z[Markers$tissue=="epithelia"])
t.test(Markers$z[Markers$tissue==goi],Markers$z[Markers$tissue=="stroma"])

length(which(!is.na(Markers$z[Markers$tissue=="stroma"])))

##Prepare Data for MarkEpi
MarkEpi.normal<-data.frame(z.normal[MarkEpi,])
if(nrow(MarkEpi.normal)<2){MarkEpi.normal<-data.frame(t(MarkEpi.normal))}
MarkEpi.primary<-data.frame(z.primary[MarkEpi,])
if(nrow(MarkEpi.primary)<2){MarkEpi.primary<-data.frame(t(MarkEpi.primary))}
if (sum(grepl("met",colnames(cstudy.ex)))>0)
{
  MarkEpi.met<-data.frame(z.met[MarkEpi,])
  if(nrow(MarkEpi.met)<2){MarkEpi.met<-data.frame(t(MarkEpi.met))}
  MarkEpi.tumor<-data.frame(rbind(MarkEpi.primary,MarkEpi.met))
  MarkEpi.list<-list(MarkEpi.normal,MarkEpi.tumor)
  MarkEpi.maxrows<- max(nrow(MarkEpi.normal),nrow(MarkEpi.tumor))
  
  MarkEpi.df <- lapply(MarkEpi.list, function(x) { x[1:MarkEpi.maxrows,] })
  MarkEpi.df<-do.call(cbind, lapply(MarkEpi.df, `[`))
  colnames(MarkEpi.df)<-c("benign", "tumor")
  
}else
{
  MarkEpi.list<-list(MarkEpi.normal,MarkEpi.primary)
  MarkEpi.maxrows<- max(nrow(MarkEpi.normal),nrow(MarkEpi.primary))
  
  MarkEpi.df <- lapply(MarkEpi.list, function(x) { x[1:MarkEpi.maxrows,] })
  MarkEpi.df<-do.call(cbind, lapply(MarkEpi.df, `[`))
  colnames(MarkEpi.df)<-c("benign", "primary")
}

##boxplot
qplot(displ, hwy, data = mpg)

df.m <- melt(MarkEpi.df)
colnames(df.m)<-c("row","status","Z")
df.m.c<-df.m[!is.na(df.m$Z),]
ggplot(data = df.m.c, aes(x=row, y=status)) + geom_boxplot(aes(fill=Z))

###########################################################
###5.1C plot goi rank order  - Requires 5.1A

##Define oncogenes
tumorsuppress<-c("PTEN", "NKX3-1")
oncogenes<-c("ERG","MYC")
goip <- c("LDAH")

##prepare data
ro.plot<-data.frame()
ro.plot<-z.full.summary
switchnames<-rownames(ro.plot)
switchnames<-gsub("C2orf43", "LDAH", switchnames)
rownames(ro.plot)<-switchnames
# if (sum(grepl("met",colnames(cstudy.ex)))>0)
# {
#   ro.plot<-z.full.summary[complete.cases(z.full.summary),]
#   ro.plot$tumors<-ro.plot$primary+ro.plot$met
#   ro.plot$ro_p<-rank(ro.plot$primary)/nrow(ro.plot)
#   ro.plot$ro_m<-rank(ro.plot$met)/nrow(ro.plot)
#   ro.plot$ro_t<-rank(ro.plot$tumor)/nrow(ro.plot)
# }else
# {
#   ro.plot<-z.full.summary[complete.cases(z.full.summary),]
#   ro.plot$ro_p<-rank(ro.plot$primary)/nrow(ro.plot)
# }

## Rank order plots
par(mar=c(2,2,1,1))

#primary
ro.plot$p_prcnt<-ro.plot$p_diff/ro.plot$p_tot
ro.plot<-ro.plot[order(ro.plot$p_prcnt, decreasing = TRUE),]
write.table(ro.plot, "ro_plot.txt")
plot(ro.plot$p_prcnt,col="cyan", ylim=c(-1,1))
text(which(row.names(ro.plot)==goip),0.75, font = 2, labels=goip, col="red", srt=90)
for(gene in oncogenes)
{
  text(which(row.names(ro.plot)==gene),0.75, labels=gene, col="green", srt=90)
} 
for(gene in tumorsuppress)
{
  text(which(row.names(ro.plot)==gene),0.3, labels=gene, col="red", srt=90)
} 

#met
ro.plot$m_prcnt<-ro.plot$m_diff/ro.plot$m_tot
ro.plot<-ro.plot[order(ro.plot$m_prcnt, decreasing = TRUE),]
plot(ro.plot$m_prcnt,col="blue", ylim=c(-1,1))
text(which(row.names(ro.plot)==goip),0.75, font = 2, labels=goip, col="red", srt=90)
for(gene in oncogenes)
{
  text(which(row.names(ro.plot)==gene),0.75, labels=gene, col="green", srt=90)
} 
for(gene in tumorsuppress)
{
  text(which(row.names(ro.plot)==gene),0.3, labels=gene, col="red", srt=90)
} 

#all tumors
ro.plot$t_prcnt<-(ro.plot$p_diff+ro.plot$m_diff)/(ro.plot$p_tot+ro.plot$m_tot)
ro.plot<-ro.plot[order(ro.plot$t_prcnt, decreasing = TRUE),]
plot(ro.plot$t_prcnt,col="black", ylim=c(-1,1))
text(which(row.names(ro.plot)==goip),0.75, font = 2, labels=goip, col="red", srt=90)
for(gene in oncogenes)
{
  text(which(row.names(ro.plot)==gene),0.75, labels=gene, col="green", srt=90)
} 
for(gene in tumorsuppress)
{
  text(which(row.names(ro.plot)==gene),0.3, labels=gene, col="red", srt=90)
} 


#ro.plot.name <-paste0(GSE,"_",GPL, "_ro_plot.txt")
#write.table(ro.plot, file = ro.plot.name, quote = FALSE, sep = "\t", row.names = TRUE, col.names = TRUE)


########################################################################################
###5.2 compare expression profile of low and high goi
###??? 1 - Most genes fall out between studies - can z threshould be lowered? Are there other ways of getting similar analysis
###??? 2 - Are there standard pathway analysis modules that you would suggest for analyzing these results (e.g., KEGG, GO)? 

##Change headers to status
colnames(cstudy.ex)<-cstudy.headers$Status_Simple
tissues<-c("primary", "met") #use "normal", "primary" or "met"
zhigh<-2.0
zlow<--2.0

ep.qz<-study.full[[GSE]][[GPL]]$GPLQNormZ
colnames(ep.qz)<-cstudy.headers$Status_Simple
ep.status<-data.frame(tissue=colnames(ep.qz))
ep.status$sig<-""
ep.status$sig[which(ep.edges[goi,]<=zlow)]<-"low"
ep.status$sig[which(ep.edges[goi,]>=zhigh)]<-"high"
ep.status$sig[which(ep.edges[goi,]>zlow & ep.edges[goi,]<zhigh)]<-"normal"
ep.status$pos<-as.numeric(as.character(rownames(ep.status)))

ep.edges<-cstudy.ex
colnames(ep.edges)<-ep.status$sig
ep.edges<-ep.edges[,ep.status$pos[ep.status$tissue%in% tissues]]
colnames(ep.edges)<-gsub("[.].*","",colnames(ep.edges))
ep.edges<-ep.edges[,which(colnames(ep.edges)=="low"|colnames(ep.edges)=="high")]
colnames(ep.edges)<-gsub("[.].*","",colnames(ep.edges))

# use instead of edges if only genes that have results in entire dataset are desired
#ep.edges.complete<-ep.edges[complete.cases(ep.edges),]

#heatmap
f<-factor(as.character(colnames(ep.edges)))
design <- model.matrix(~f)
fit <- eBayes(lmFit(ep.edges,design))

selected <- p.adjust(fit$p.value[, 2], method="bonferroni") <0.05
exSel <- data.matrix(ep.edges[selected, ])
exSel <- exSel[which(rowSums(exSel, na.rm = TRUE)!=0),]
dist_exSel <- dist(exSel)
df_exSel <- data.frame(t(combn(rownames(exSel),2)), as.numeric(dist_exSel))
colnames(df_exSel)<-c("first","second","distance")

# coregs<-which(selected==TRUE)
# write.table(coregs, file = "coregs.txt")
# write.table(selected, file = "allgenes.txt")

#cleans out genes which have a distance of <NA> - i.e. those never seen on the same chip
while(length(df_exSel$first[is.na(df_exSel$distance)])>=1)
{
  firstcount<-count(df_exSel$first[is.na(df_exSel$distance)])
  secondcount<-count(df_exSel$second[is.na(df_exSel$distance)])
  counts<-merge(firstcount,secondcount, by="x", all = TRUE)
  rownames(counts)<-counts$x
  counts$x<-NULL
  counts$sum <-rowSums(counts, na.rm = TRUE)
  badgene <- rownames(counts[which.max(counts$sum),])
  exSel<-exSel[rownames(exSel)!=badgene,]
  dist_exSel <- dist(exSel)
  df_exSel <- data.frame(t(combn(rownames(exSel),2)), as.numeric(dist_exSel))
  colnames(df_exSel)<-c("first","second","distance")
}

heatmap.2(exSel, col=redgreen(75), na.color="grey", scale="none",key=TRUE, symkey=FALSE, density.info="none", trace="none", cexRow=0.75, cexCol=0.5, lwid=c(0.5,2))
genelist<-names(selected[which(selected==TRUE)])
write.table(genelist, file = "genelist.txt", quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)


goi.z.edges <-paste0(prefix,"_",goi,"_",tissue,"_edges.txt")
write.table(goi.df, file = goi.rankplot.name, quote = FALSE, sep = "\t", row.names = TRUE, col.names = TRUE)

#################################################################
###5.3 survival curves of goi
##Change headers to status and initiate
colnames(cstudy.ex)<-cstudy.headers$Status_Simple
study.survival <- cstudy.ex

##change to correct headers
colnames(study.survival)<-cstudy.headers$ID_Simple

study.survival<-study.survival[rownames(study.survival)==goi,]
study.survival["ID_Simple",]<-colnames(study.survival)
study.survival["Status",]<-cstudy.headers$Status_Simple
study.survival<-data.frame(t(study.survival))


study.survival$C2orf43<-as.vector(study.survival$C2orf43)
study.survival$low<-as.numeric(study.survival$C2orf43)<=-2
study.survival$ID_Simple<-as.character(study.survival$ID_Simple)
study.survival$ID_Simple<-gsub(" ","",study.survival$ID_Simple)

cdata.survival<-cstudy.cdata
cdata.survival$ID_Simple<-as.character(cdata.survival$ID_Simple)

survdata<-merge(cdata.survival, study.survival, by="ID_Simple", all = TRUE)
survdata$Surv_Time<-as.numeric((as.character(survdata$Surv_Time)))
survdata$SurvObj <- with(survdata, Surv(Surv_Time, Surv_Event == 1))
survdata_primary <- survdata[survdata$Status=="primary",]
survdata_met <- survdata[survdata$Status=="met",]

objNpsurv_primary <- npsurv(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata_primary)
survdiff(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata_primary)
survplot(objNpsurv_primary)

objNpsurv_met <- npsurv(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata_met)
survdiff(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata_met)
survplot(objNpsurv_met) 

objNpsurv <- npsurv(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata)
survdiff(formula = Surv(Surv_Time,Surv_Event == 1) ~ low, data = survdata)
survplot(objNpsurv)


write.table(survdata, "survdata.txt")

#########################################################


###########################################################
########################END################################
###########################################################
